idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/ReplaceCallFixFactories.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.quickfix.fixes
+
+import com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.idea.KotlinBundle
+import org.jetbrains.kotlin.idea.fir.api.applicator.HLApplicatorInput
+import org.jetbrains.kotlin.idea.fir.api.applicator.applicatorByQuickFix
+import org.jetbrains.kotlin.idea.fir.api.fixes.HLApplicatorWithTargetAndInput
+import org.jetbrains.kotlin.idea.fir.api.fixes.diagnosticFixFactory
+import org.jetbrains.kotlin.idea.fir.api.fixes.withInput
+import org.jetbrains.kotlin.idea.frontend.api.fir.diagnostics.KtFirDiagnostic
+import org.jetbrains.kotlin.idea.frontend.api.types.KtTypeNullability
+import org.jetbrains.kotlin.idea.frontend.api.types.KtTypeWithNullability
+import org.jetbrains.kotlin.idea.quickfix.ReplaceCallFix
+import org.jetbrains.kotlin.idea.quickfix.ReplaceImplicitReceiverCallFix
+import org.jetbrains.kotlin.idea.quickfix.ReplaceWithSafeCallFix
+import org.jetbrains.kotlin.psi.*
+
+object ReplaceCallFixFactories {
+    private val replaceWithSafeCallFixApplicator =
+        applicatorByQuickFix<KtExpression, Input, ReplaceCallFix>(getFamilyName = { KotlinBundle.message("replace.with.safe.call") },
+                                                                  isApplicableByPsi = { psi -> psi is KtDotQualifiedExpression }) { psi, input ->
+            ReplaceWithSafeCallFix(psi as KtDotQualifiedExpression, input.notNullNeeded)
+        }
+
+    private val replaceImplicitReceiverCallFixApplicator =
+        applicatorByQuickFix<KtExpression, Input, ReplaceImplicitReceiverCallFix>(getFamilyName = { KotlinBundle.message("replace.with.safe.this.call") }) { psi, input ->
+            ReplaceImplicitReceiverCallFix(psi, input.notNullNeeded)
+        }
+
+    class Input(val notNullNeeded: Boolean) : HLApplicatorInput
+
+    val unsafeCallFactory =
+        diagnosticFixFactory<PsiElement, KtFirDiagnostic.UnsafeCall, KtExpression, Input> { diagnostic ->
+            fun KtExpression.shouldHaveNotNullType(): Boolean {
+                // This function is used to determine if we may need to add an elvis operator after the safe call. For example, to replace
+                // `s.length` in `1 + s.length` with a safe call, it should be replaced with `(s.length ?: <caret>)`.
+                // NOTE: Even though we could, we don't check to see if the call is an argument and if the parameter type is nullable.
+                // FE1.0 doesn't either (see ReplaceWithSafeCallFixFactory).
+                val type = when (val parent = parent) {
+                    is KtBinaryExpression -> parent.left?.getKtType()
+                    is KtProperty -> {
```

Comment:
Should we check the same for `KtNamedFunction` for cases like `fun foo(s: String?) = s.length`?

Reply:
This isn't checked in FE1.0 (see idea/src/org/jetbrains/kotlin/idea/quickfix/ReplaceCallFixUtils.kt). E.g., for `fun foo(s: String?): Int = s.length` the FE1.0 quickfix does _not_ replace with `(s?.length ?: <caret>)`

In general, should we try to match FE1.0 capabilities exactly, or is it ok to do more?

Reply:
Let's do more :) If we can do better than in FE10, why not to?

Reply:
And there is a function in  KtAnalysisSession `PsiElement.getExpectedType`: let's use it for determining expected type

Reply:
@darthorimar I _think_ I was able to follow your suggestion, see 4th commit onwards. Let me know what you think.

I needed to add support for FIR-specific "after" files in the quickfix tests since FE1.0 behavior is different for some of them (I adapted @tgeng's changes [here](https://github.com/JetBrains/kotlin/pull/4185/files#diff-4a0fbd9ab76ad5ff7f4d8b274b4fa7d6b5d67ec47514cefa07653805ffa46bea) 🙂). In all differences, the FIR behavior is strictly better. 🙂